### PR TITLE
PID implementation inside mujoco-py

### PR DESCRIPTION
**Pull request by bayesian**
Expose 5 new parameters for full PID with clamp/smoothing. 

test plans:
1. verified the correctness, when client do not specify new parameters, PID behaves like P
2. for mujoco_shadow_hand test, with global I/D settings, position errors after 100 steps become smaller.
3. unit tests all with inverted pendulum

### CODE REVIEW

File: mujoco_py/mjpid.pyx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    if m.nuserdata < m.nu * NUM_USER_DATA_PER_ACT:
+        raise Exception('nuserdata is not set large enough to store PID internal states')
+
+    for i in range(m.nuserdata):
+        d.userdata[i] = 0.0
+
+    for i in range(m.nu):
+        m.actuator_gaintype[i] = const.GAIN_USER
+        m.actuator_biastype[i] = const.BIAS_USER
```
Comment (MillionIntegrals) [inline mujoco_py/mjpid.pyx]: I don't think we should override that manually for the actuators here. That should be specified in the XML file.

File: mujoco_py/mjpid.pyx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        # if user does not set, will chose default settings.
+        # for kp, it tries to use m.actuator_gainprm[i][0]
+        if m.actuator_user[i][PROPORTIONAL_GAIN] <= 0.0:
```
Comment (MillionIntegrals) [inline mujoco_py/mjpid.pyx]: I know it's not that important, but why not use gainprm for this purpose?

**Comment by arthurpetron** [comment]
No! They are that way for a reason!

File: mujoco_py/mjpid.pyx (hunk 1 above)
Comment (bayesian) [reply mujoco_py/mjpid.pyx]: if we set xml type to "user", it forces everyone to switch to new PID; we certainly could, I shall choose a setting to be exactly the same as previous position control.  

I was thinking about "opt in", only if user set_pid_control, to turn on the behavior

Comment (MillionIntegrals) [reply mujoco_py/mjpid.pyx]: I also would like it to be opt-in, but on the XML level - then we can have multiple hands, one with PID Controller, one without. We can have even these two hand in the same simulation. Or we can have kuka arm with P controllers and a shadow hand with PID controllers. The more things we allow to configure via XML the better for flexibility.

File: mujoco_py/mjpid.pyx (hunk 2 above)
Comment (MillionIntegrals) [reply mujoco_py/mjpid.pyx]: I also see this kind of as a good thing. If we keep the `kp` at the same position, just by changing the `gaintype` from `user` to `position`, we can change between the P and PID controller with the same P value. Pretty cool, isn't it?

File: mujoco_py/mjpid.pyx [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    cdef double error = d.ctrl[id] - d.actuator_length[id]
+    cdef int NGAIN = int(const.NGAIN)
+
+    cdef double Kp = m.actuator_gainprm[id * NGAIN + PROPORTIONAL_GAIN]
+    cdef double error_deadband = m.actuator_gainprm[id * NGAIN + ERROR_DEADBAND]
+    cdef double integral_max_clamp = m.actuator_gainprm[id * NGAIN + INTEGRAL_MAX_CLAMP]
+    cdef double integral_time_const = m.actuator_gainprm[id * NGAIN + INTEGRAL_TIME_CONSTANT]
+    cdef double derivative_gain_smoothing = \
+        m.actuator_gainprm[id * NGAIN  + DERIVATIVE_GAIN_SMOOTHING]
+    cdef double derivate_time_const = m.actuator_gainprm[id * NGAIN + DERIVATIVE_TIME_CONSTANT]
```
Comment (arthurpetron) [inline mujoco_py/mjpid.pyx]: By DERIVATIVE_TIME_CONSTANT do you actually mean DERIVATIVE_TIME_CONSTANT_ID? Same comment for lines 44 - 50 and lines 20 - 25 and lines 29 - 32. What is NGAIN? Is it also an ID?

Comment (bayesian) [reply mujoco_py/mjpid.pyx]: DERIVATIVE_TIME_CONSTANT is the index for the field, NGAIN is the total number of parameters

Comment (bayesian) [reply mujoco_py/mjpid.pyx]: updated to IDX_

**Comment by Jendker** [comment]
Thanks for a great PR! 

May I ask you if it is working fine with shadow robot hand? I was trying to use this PID control, but the finger control does not work anymore, the fingers don't move a bit.

I tried also the P control equivalent case based on the commited pid_test, but the fingers are still immobile. If the shadow robot should work I will open a separate issue.

**Comment by bayesian** [comment]
yes, it works for us on our shadow hand simulation and policy training, it is also backward compatible.  you need to change xml and call enable_pid, like in the test_pid.py in the PR.

**Comment by Jendker** [comment]
Thank you for your quick answer! But are you always setting up both gaintype and biastype to "user"? If I define gaintype="user" the fingers stop any movement.

**Comment by bayesian** [comment]
you have to set biastype="user" and gaintype="user", as our pid implementation use user-defined callback: mjcb_act_gain and mjcb_act_bias to implement full PID control.

you can read more about mujoco actuator model here: http://www.mujoco.org/book/XMLreference.html#actuator

change your xml here: https://github.com/Jendker/pid_shadow_robot/blob/master/DAPG_relocate_P_control.xml#L80

**Comment by brinij** [comment]
Hi,
I would also like to have shadow hand mujoco simulation with PID controller.
My only question now is: are you using the same hand model that is in gym environments: robotics/assets/hand/shared.xml or some other? Because the joint names are not quite the same and I am wondering weather this PID parameters will work on other model as well?
Thank you in advance!

**Comment by bayesian** [comment]
you are free to choose the PID settings, the example settings are from one of my experiments. our hand model used for sim2real is a bit different from gym version.

**Pull request merged**